File: Engine/UI/RootWidget.py
# File name: RootWidget.py
# This file is part of: pyengine
#
# LICENSE
#
# The contents of this file are subject to the Mozilla Public License
# Version 1.1 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS"
# basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
# the License for the specific language governing rights and limitations
# under the License.
#
# Alternatively, the contents of this file may be used under the terms
# of the GNU General Public license (the  "GPL License"), in which case
# the provisions of GPL License are applicable instead of those above.
#
# FEEDBACK & QUESTIONS
#
# For feedback and questions about pyengine please e-mail one of the
# authors named in the AUTHORS file.
########################################################################
from __future__ import unicode_literals, print_function, division
from our_future import *

__all__ = ["RootWidget"]

import cairo
import copy

import Engine.CEngine.Pango as Pango

import CSS.Minilanguage

from Engine.CEngine import key, motion
from WidgetBase import AbstractWidget, WidgetContainer
from LayerWidget import LayerWidget, DesktopLayer, WindowLayer, PopupLayer
from Flags import *

from Engine.GL import make_pot

class RootWidget(AbstractWidget, WidgetContainer):
    """
    This widget implements dispatching of events and can work as a parent
    for other widgets. It also handles focus and mouse capturing. For this
    it respects the *Focusable* flag, which must be set by a widget which
    wants to recieve the focus.
    """

    keymap = {
        (key.Left, 0): lambda self: self.dispatch_caret_motion(
            motion.Left, motion.StepChar),
        (key.Right, 0): lambda self: self.dispatch_caret_motion(
            motion.Right, motion.StepChar),
        (key.Home, 0): lambda self: self.dispatch_caret_motion(
            motion.Begin, motion.StepChar),
        (key.End, 0): lambda self: self.dispatch_caret_motion(
            motion.End, motion.StepChar)
        }

    def __init__(self, **kwargs):
        self._theme = None
        self._invalidated_rects = []  # needed during initialization
        super(RootWidget, self).__init__(**kwargs)
        self._rootwidget = self
        self._mouse_capture = None
        self._mouse_capture_button = 0
        self._focused = None
        self._flags = frozenset()
        self._child_classes = LayerWidget
        self._popuplayer = PopupLayer(self)
        self._windowlayer = WindowLayer(self)
        self._desktoplayer = DesktopLayer(self)
        self._old_hit_chain = frozenset()
        self._cairo_surface = None
        self._cairo = None
        self._pango = None
        self._resized = False
        self._invalidated_rects = []  # discard init values, they're incorrect
                                     # anyways
        self.surface_dirty = False

    def _find_key_event_target(self):
        if self._focused and self._focused.RootWidget is not self:
            self._focused = None
        return self._focused or self

    def _map_mouse_event(self, x, y, hitchain=None):
        target = self._mouse_capture or \
            ((hitchain[0] if hitchain is not False else None)
             if hitchain is not None else self._hit_test((x, y)))
        if target is None:
            return None, x, y
        return (target, x - target.AbsoluteRect.Left, y - target.AbsoluteRect.Top)

    def _update_hover_state(self, hitchain=None, p=None):
        if hitchain is None:
            hitchain = self._hit_test_with_chain(p)
        if hitchain is False:
            hitchain = []
        hitchain = frozenset(hitchain)
        old_hit_chain = self._old_hit_chain
        if hitchain != old_hit_chain:
            for non_hovered in old_hit_chain - hitchain:
                non_hovered.IsHovered = False
                non_hovered.on_mouse_leave()

            for hovered in hitchain - old_hit_chain:
                hovered.IsHovered = True
                hovered.on_mouse_enter()

            self._old_hit_chain = hitchain

    def _focus_and_capture(self, hitchain, button):
        target = None
        for candidate in hitchain:
            if Focusable in candidate._flags:
                target = candidate
                break
        else:
            return

        if self._focused is not None:
            self._focused._is_focused = False
            self._focused._invalidate_computed_style()
        self._focused = target
        target.IsFocused = True

        self._mouse_capture = target
        self._mouse_capture_button = button

    def _recreate_cairo_context(self, width, height):
        self._cairo_surface = cairo.ImageSurface(
            cairo.FORMAT_ARGB32,
            width, height)
        self._cairo = cairo.Context(self._cairo_surface)
        self._pango = Pango.PangoCairo.create_context(self._cairo)
        self.update_rendering_context()

    def _require_cairo_context(self):
        myrect = self.AbsoluteRect
        w, h = myrect.Width, myrect.Height

        if not (self._cairo_surface and
                w == self._cairo_surface.get_width() and
                h == self._cairo_surface.get_height()):
            self._recreate_cairo_context(w, h)

    def invalidate_rect(self, rect):
        self._invalidated_rects.append(copy.copy(rect))

    def do_align(self):
        assert len(self) == 3
        myrect = self.AbsoluteRect
        self._desktoplayer.AbsoluteRect = myrect
        self._windowlayer.AbsoluteRect = myrect
        self._popuplayer.AbsoluteRect = myrect

        self._require_cairo_context()
        self._invalidated_rects = []
        self._resized = True

    def dispatch_key_down(self, key, modifiers):
        target = self._find_key_event_target()

        try:
            handler = self.keymap[(key, modifiers)]
        except KeyError:
            pass
        else:
            if handler(self):
                return

        handled = target.on_key_down(key, modifiers)
        parent = target.Parent
        while parent and not handled:
            handled = parent.on_key_down(key, modifiers)
            parent = parent.Parent

    def dispatch_key_up(self, *args):
        target = self._find_key_event_target()
        handled = target.on_key_up(*args)
        parent = target.Parent
        while parent and not handled:
            handled = parent.on_key_up(*args)
            parent = parent.Parent

    def dispatch_mouse_down(self, x, y, button, modifiers):
        if self._mouse_capture is None:
            hitchain = self._hit_test_with_chain((x, y))
        else:
            hitchain = None
        target, x, y = self._map_mouse_event(x, y, hitchain)
        if target:
            target.on_mouse_down(x, y, button, modifiers)
        if self._mouse_capture is None:
            self._focus_and_capture(hitchain, button)

    def dispatch_mouse_click(self, x, y, button, modifiers, nth):
        target, cx, cy = self._map_mouse_event(x, y)
        if target:
            target.on_mouse_click(cx, cy, button, modifiers, nth)

    def dispatch_mouse_move(self, x, y, dx, dy, button, modifiers):
        if self._mouse_capture is None:
            hitchain = self._hit_test_with_chain((x, y))
        else:
            hitchain = None
        target, x, y = self._map_mouse_event(x, y, hitchain)
        if target:
            target.on_mouse_move(x, y, dx, dy, button, modifiers)

        if self._mouse_capture is None:
            self._update_hover_state(hitchain)

    def dispatch_mouse_up(self, x, y, button, modifiers):
        target, cx, cy = self._map_mouse_event(x, y)
        if target:
            target.on_mouse_up(cx, cy, button, modifiers)
        if target is self._mouse_capture and button & self._mouse_capture_button:
            self._mouse_capture = None
            self._mouse_capture_button = 0
            self._update_hover_state(p=(x,y))

    def dispatch_scroll(self, x, y, scrollX, scrollY):
        target, x, y = self._map_mouse_event(x, y)
        if target:
            target.on_scroll(scrollX, scrollY)

    def dispatch_text_input(self, text):
        target = self._find_key_event_target()
        target.on_text_input(text)

    def dispatch_caret_motion(self, direction, step):
        target = self._find_key_event_target()
        handled = target.on_caret_motion(direction, step)
        parent = target.Parent
        while not handled and parent:
            handled = parent.on_caret_motion(direction, step)
            parent = parent.Parent

    def dispatch_caret_motion_select(self, direction, step):
        target = self._find_key_event_target()
        handled = target.on_caret_motion_select(direction, step)
        parent = target.Parent
        while not handled and parent:
            handled = parent.on_caret_motion_select(direction, step)
            parent = parent.Parent
        return handled

    def realign(self):
        super(RootWidget, self).realign()
        for child in self:
            child.realign()

    def clear_cairo_surface(self):
        ctx = self._cairo
        ctx.set_source_rgba(0., 0., 0., 0.)
        ctx.set_operator(cairo.OPERATOR_SOURCE)
        ctx.paint()
        ctx.set_operator(cairo.OPERATOR_OVER)
        ctx.set_line_cap(cairo.LINE_CAP_SQUARE)

    def _setup_clipping(self):
        ctx = self._cairo
        for rect in self._invalidated_rects:
            ctx.rectangle(*rect.XYWH)
        ctx.clip()

    def render(self):
        self.realign()
        self.surface_dirty = False
        if not (self._invalidated_rects or self._resized):
            # no dirty regions, nothing to do \o/
            return
        if not self._resized:
            # set up clipping regions
            self._setup_clipping()
        self.clear_cairo_surface()
        self._desktoplayer.render()
        self._windowlayer.render()
        self._popuplayer.render()
        self._invalidated_rects = []
        self._resized = False
        self._cairo.reset_clip()
        self.surface_dirty = True

    def update(self, timedelta):
        for child in self:
            child.update(timedelta)

    @property
    def WindowLayer(self):
        return self._windowlayer

    @property
    def DesktopLayer(self):
        return self._desktoplayer

    @property
    def PopupLayer(self):
        return self._popuplayer

    @property
    def Parent(self):
        return None

    @property
    def RootWidget(self):
        return self

    @property
    def Theme(self):
        return self._theme

    @Theme.setter
    def Theme(self, theme):
        if theme == self._theme:
            return
        self._theme = theme
        for widget in self.tree_depth_first():
            widget._theme_changed()

CSS.Minilanguage.ElementNames().register_widget_class(RootWidget, "Root")

File: CEngine/UI/Widgets/RootWidget.cpp
/**********************************************************************
File name: RootWidget.cpp
This file is part of: Pythonic Engine

LICENSE

The contents of this file are subject to the Mozilla Public License
Version 1.1 (the "License"); you may not use this file except in
compliance with the License. You may obtain a copy of the License at
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS"
basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
License for the specific language governing rights and limitations under
the License.

Alternatively, the contents of this file may be used under the terms of
the GNU General Public license (the  "GPL License"), in which case  the
provisions of GPL License are applicable instead of those above.

FEEDBACK & QUESTIONS

For feedback and questions about pyengine please e-mail one of the
authors named in the AUTHORS file.
**********************************************************************/
#include "RootWidget.hpp"

#include <stdexcept>

#include <pango/pangocairo.h>

#include "CEngine/Misc/SetOperations.hpp"

namespace PyEngine { namespace UI {

/* PyEngine::UI::RootWidget */

RootWidget::RootWidget():
    ParentWidget(),
    EventSink(),
    _theme(new Theme()),
    _invalidated_rect(),
    _mouse_capture(nullptr),
    _mouse_capture_button(0),
    _focused(nullptr),
    _old_hit_chain(),
    _old_focus_chain(),
    _cairo_ctx(nullptr),
    _cairo_surface(nullptr),
    _pango_ctx(nullptr),
    _drag_controller(nullptr),
    _cursor(0, 0),
    _surface_dirty(),
    _desktop_layer(new DesktopLayer()),
    _window_layer(new WindowLayer()),
    _popup_layer(new PopupLayer())
{
    _invalidated_rect = NotARect;
    add(_desktop_layer);
    add(_window_layer);
    add(_popup_layer);
}

RootWidget::~RootWidget()
{

}

void RootWidget::_capture(WidgetPtr capturee, unsigned int button)
{
    if (!(capturee->flags() & WidgetFlag::CAPTURE_MOUSE)) {
        return;
    }
    _mouse_capture = capturee;
    _mouse_capture_button = button;
}

void RootWidget::_clear_cairo_surface()
{
    cairo_set_source_rgba(_cairo_ctx, 0, 0, 0, 0);
    cairo_set_operator(_cairo_ctx, CAIRO_OPERATOR_SOURCE);
    cairo_paint(_cairo_ctx);
    cairo_set_operator(_cairo_ctx, CAIRO_OPERATOR_OVER);
    cairo_set_line_cap(_cairo_ctx, CAIRO_LINE_CAP_SQUARE);
}

WidgetPtr RootWidget::_find_key_event_target()
{
    if (_focused && (_focused->get_root() != this)) {
        _focused = nullptr;
    }
    return (_focused ? _focused : get_root());
}

WidgetPtr RootWidget::_find_mouse_event_target(
    int x, int y, HitChain *chain)
{
    if (_mouse_capture) {
        return _mouse_capture;
    }

    if (chain) {
        if (chain->size() == 0) {
            return WidgetPtr();
        } else {
            return chain->front();
        }
    }

    return hittest(Point(x, y));
}

void RootWidget::_focus(HitChain *chain)
{
    if (chain == nullptr) {
        return;
    }

    WidgetPtr target = nullptr;
    HitChain::iterator target_iter = chain->end();
    for (auto it = chain->begin(); it != chain->end(); it++)
    {
        target = *it;
        if (target->flags() & WidgetFlag::FOCUSABLE) {
            target_iter = it;
            break;
        }
    }
    if (target_iter == chain->end()) {
        return;
    }

    if (_focused) {
        _focused->state().reset(Focused);
    }
    _focused = target;
    target->state().set(Focused);

    target_iter++;
    HitSet new_set(target_iter, chain->end());

    auto diff = set_difference_iterator(new_set, _old_focus_chain);
    for (; diff != decltype(diff)(); ++diff) {
        const WidgetPtr &focused = *diff;
        focused->state().set(HasFocusedChild);
        focused->invalidate_computed_style();
    }

    diff = set_difference_iterator(_old_focus_chain, new_set);
    for (; diff != decltype(diff)(); ++diff) {
        const WidgetPtr &unfocused = *diff;
        unfocused->state().reset(HasFocusedChild);
        unfocused->invalidate_computed_style();
    }

    std::swap(_old_focus_chain, new_set);
}

void RootWidget::_recreate_cairo_surface(unsigned int width,
        unsigned int height)
{
    if (_cairo_surface) {
        cairo_surface_destroy(_cairo_surface);
    }
    if (_cairo_ctx) {
        cairo_destroy(_cairo_ctx);
    }
    if (_pango_ctx) {
        g_object_unref(_pango_ctx);
    }

    _cairo_surface = cairo_image_surface_create(
        CAIRO_FORMAT_ARGB32,
        width,
        height);
    _cairo_ctx = cairo_create(_cairo_surface);
    _pango_ctx = pango_cairo_create_context(_cairo_ctx);

    invalidate_context();
}

void RootWidget::_require_cairo_context()
{
    coord_int_t w = absolute_rect().get_width();
    coord_int_t h = absolute_rect().get_height();

    assert(w >= 0);
    assert(h >= 0);

    if ((_cairo_surface == nullptr) ||
        (cairo_image_surface_get_width(_cairo_surface) != w) ||
        (cairo_image_surface_get_height(_cairo_surface) != h))
    {
        _recreate_cairo_surface(w, h);
    }
}

void RootWidget::_setup_clipping()
{
    cairo_rectangle(
        _cairo_ctx,
        _invalidated_rect.get_x(),
        _invalidated_rect.get_y(),
        _invalidated_rect.get_width(),
        _invalidated_rect.get_height());
    cairo_clip(_cairo_ctx);
}

void RootWidget::_update_hover_state(HitChain *chain)
{
    if (chain == nullptr) {
        std::unique_ptr<HitChain> tmp_chain(new HitChain());
        _update_hover_state(tmp_chain.get());
        return;
    }

    HitSet new_set(chain->begin(), chain->end());
    auto diff = set_difference_iterator(new_set, _old_hit_chain);
    for (; diff != decltype(diff)(); ++diff) {
        const WidgetPtr &hovered = *diff;
        hovered->state().set(Hovered);
        hovered->invalidate_computed_style();
    }

    diff = set_difference_iterator(_old_hit_chain, new_set);
    for (; diff != decltype(diff)(); ++diff) {
        const WidgetPtr &hovered = *diff;
        hovered->state().reset(Hovered);
        hovered->invalidate_computed_style();
    }

    std::swap(_old_hit_chain, new_set);
}

void RootWidget::do_align()
{
    for (auto& child: *this) {
        child->absolute_rect() = absolute_rect();
        child->invalidate_alignment();
    }

    _require_cairo_context();
    _invalidated_rect = NotARect;
}

cairo_t* RootWidget::get_cairo_context()
{
    return _cairo_ctx;
}

PangoContext* RootWidget::get_pango_context()
{
    return _pango_ctx;
}

RootPtr RootWidget::get_root()
{
    return this;
}

ThemePtr RootWidget::get_theme()
{
    return _theme;
}

bool RootWidget::is_element(const std::string &name) const
{
    return (name == "root");
}

void RootWidget::render()
{
    realign();
    _surface_dirty = false;
    bool full_repaint = (_invalidated_rect == NotARect);
    if ((!full_repaint) && (_invalidated_rect.area() == 0))
    {
        return;
    }

    if (!full_repaint) {
        _setup_clipping();
    }

    _clear_cairo_surface();
    _desktop_layer->render();
    _window_layer->render();
    _popup_layer->render();
    _invalidated_rect = Rect(0, 0, 0, 0);

    cairo_reset_clip(_cairo_ctx);
    _surface_dirty = true;
}

void RootWidget::set_theme(const ThemePtr &theme)
{
    _theme = theme;
}

void RootWidget::focus(const WidgetPtr &ref)
{
    HitChain chain;
    iter_upwards(ref, std::back_inserter(chain));
    _focus(&chain);
}

void RootWidget::invalidate_rect(const Rect &ref)
{
    _invalidated_rect.extend(ref);
}

void RootWidget::release_capture()
{
    _mouse_capture = nullptr;
    _mouse_capture_button = 0;
}

void RootWidget::start_drag(
    std::unique_ptr<DragControllerBase> &&controller)
{
    if (_drag_controller) {
        throw std::invalid_argument("Dragging is already in progress.");
    }
    _drag_controller = std::move(controller);
    release_capture();
}

void RootWidget::stop_drag()
{
    _drag_controller = nullptr;
    release_capture();
}

/* event dispatching */

void RootWidget::dispatch_caret_motion(
    CaretMotionDirection direction,
    CaretMotionStep step,
    bool select)
{
    WidgetPtr target = _find_key_event_target();
    bool handled = false;
    while (target && !handled) {
        handled = target->ev_caret_motion(direction, step, select);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_hide()
{

}

void RootWidget::dispatch_key_down(Key::Key key,
                                   KeyModifiers modifiers)
{
    if (_drag_controller) {
        _drag_controller->ev_key_down(key, modifiers);
        return;
    }

    /* TODO: dispatch caret motion (and other global bindings) */

    WidgetPtr target = _find_key_event_target();
    bool handled = false;
    while (target && !handled) {
        handled = target->ev_key_down(key, modifiers);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_key_up(Key::Key key,
                                 KeyModifiers modifiers)
{
    if (_drag_controller) {
        _drag_controller->ev_key_up(key, modifiers);
        return;
    }

    WidgetPtr target = _find_key_event_target();
    bool handled = false;
    while (target && !handled) {
        handled = target->ev_key_up(key, modifiers);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_mouse_click(
    int x, int y,
    MouseButton button,
    KeyModifiers modifiers,
    unsigned int nth)
{
    if (_drag_controller) {
        return;
    }

    WidgetPtr target = _find_mouse_event_target(x, y);
    bool handled = false;
    while (target && !handled) {
        const Point &translated = target->absolute_to_client(x, y);

        handled = target->ev_mouse_click(translated.get_x(),
                                         translated.get_y(),
                                         button, modifiers, nth);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_mouse_down(
    int x, int y,
    MouseButton button,
    KeyModifiers modifiers)
{
    if (_drag_controller) {
        _drag_controller->ev_mouse_down(x, y, button, modifiers);
        return;
    }

    std::unique_ptr<HitChain> hitchain;
    if (!_mouse_capture) {
        hitchain = std::unique_ptr<HitChain>(new HitChain());
        hittest_with_chain(Point(x, y), *hitchain.get());
    }

    _focus(hitchain.get());

    WidgetPtr target;
    target = _find_mouse_event_target(x, y, hitchain.get());

    bool handled = false;
    while (target && !handled) {
        const Point &translated = target->absolute_to_client(x, y);

        handled = target->ev_mouse_down(translated.get_x(),
                                        translated.get_y(),
                                        button, modifiers);
        if (handled) {
            // we must break here as we need the actual target for
            // capturing
            break;
        }
        target = target->get_parent();
    }

    if (handled && !_mouse_capture) {
        _capture(target, button);
    }
}

void RootWidget::dispatch_mouse_move(
    int x, int y,
    int dx, int dy,
    MouseButtons buttons,
    KeyModifiers modifiers)
{
    _cursor = std::move(Point(x, y));
    if (_drag_controller) {
        _drag_controller->ev_mouse_move(x, y, buttons, modifiers);
        return;
    }

    std::unique_ptr<HitChain> hitchain;
    if (!_mouse_capture) {
        hitchain = std::unique_ptr<HitChain>(new HitChain());
        hittest_with_chain(Point(x, y), *hitchain.get());
    }

    WidgetPtr target;
    target = _find_mouse_event_target(x, y, hitchain.get());

    bool handled = false;
    while (target && !handled) {
        const Point &translated = target->absolute_to_client(x, y);

        handled = target->ev_mouse_move(translated.get_x(),
                                        translated.get_y(),
                                        dx, dy,
                                        buttons, modifiers);
        if (_mouse_capture) {
            // no hover update and no propagation upwards on captured
            // mouse!
            return;
        }
        if (handled) {
            break;
        }
        target = target->get_parent();
    }

    _update_hover_state(hitchain.get());
}

void RootWidget::dispatch_mouse_up(
    int x, int y,
    MouseButton button,
    KeyModifiers modifiers)
{
    if (_drag_controller) {
        _drag_controller->ev_mouse_up(x, y, button, modifiers);
        return;
    }

    WidgetPtr target;
    target = _find_mouse_event_target(x, y);

    bool handled = false;
    while (target && !handled) {
        const Point &translated = target->absolute_to_client(x, y);

        handled = target->ev_mouse_up(translated.get_x(),
                                      translated.get_y(),
                                      button, modifiers);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_resize(
    unsigned int w,
    unsigned int h)
{
    absolute_rect() = Rect(0, 0, w, h);
    invalidate_alignment();
}

void RootWidget::dispatch_scroll(
    int x, int y,
    int scrollx, int scrolly)
{
    WidgetPtr target;
    target = _find_mouse_event_target(x, y);

    bool handled = false;
    while (target && !handled) {
        handled = target->ev_scroll(scrollx, scrolly);
        if (handled) {
            break;
        }
        target = target->get_parent();
    }
}

void RootWidget::dispatch_show()
{

}

void RootWidget::dispatch_text_input(const char *text)
{
    if (_drag_controller) {
        return;
    }

    WidgetPtr target = _find_key_event_target();
    if (target) {
        target->ev_text_input(text);
    }
}

void RootWidget::dispatch_wm_quit()
{

}

void RootWidget::frame_synced()
{

}

void RootWidget::frame_unsynced(TimeFloat deltaT)
{

}

}
}
